---
title: "Tutorial 2: Train the network model  "
author: "Chixiang Chen"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{DRDNetProf}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '/Users/chixiang.chen/OneDrive - University of Maryland School of Medicine/phd_in_psu/research/Prof._Wu/Network/tissue/blood_vessel/')
```

**DRDNetProf** is a statistical learning algorithm for recover disease risk-associated pseudo-dynamic networks (DRDNet) from steady-state data. It incorporates risk prediction model, a varying coefficient model, multiple ordinary differential equations, and group lasso estimation to learn a series of networks.This tutorial will provide detailed information for the second step of implementing the [DRDNet](https://academic.oup.com/bioinformatics/article-abstract/38/9/2481/6537533). If the first step has not been done yet, please visit ?? 

```{r}
library(np)
library(splines2)
library(grpreg)
library(Matrix)
```

# Prepare the data for training the network
Before recovering the network, the users need to preprocess the gene expression data. This procedure will involve following two steps: gene expression preprocess and gene selection
## Filtering and transformation
This step involves deleting gene with all zeros, log transformation, and choose genes without zero values. The last filtering step is to stabilize the algorithm, which is not required in general.
```{r,message=FALSE}
#read the data
pheno_together<-read.csv("pheno_together.csv") # phenotype data
data_vessel<-read.csv("data_vessel.csv") # gene expression data

rownames(data_vessel)<-data_vessel[,1]
data_vessel<-data_vessel[,-1]
dim(data_vessel)

#delete gene with all zeros
row_mean<-apply(data_vessel,1, mean)
data_vessel<-data_vessel[row_mean>0,]
dim(data_vessel)

#let the gene expression equal to zero if it's less than 1 level:
data_vessel[data_vessel<=1]<-0
dim(data_vessel)
which(data_vessel<=1 & data_vessel>0)

#count zeros in each row:
zero_row<-apply(data_vessel,1,function (x) length(which(x==0)))
summary(zero_row)

#do log transformation
data_vessel[data_vessel!=0]<-log(data_vessel[data_vessel!=0])
dim(data_vessel)

#set threshold about zeros:
data_vessel<-data_vessel[zero_row<=0,]
dim(data_vessel)

write.csv(data_vessel,"data_vessel_processed.csv",row.names =F)
```

The reprocessed data can be directly loaded by
```{r,message=FALSE}
data_vessel<-read.csv("data_vessel_processed.csv") # read preprocessed gene expression data
```

## Gene selection (screening)
first create the data for screening
```{r,message=FALSE}
vessel_subject_id<-substring(colnames(data_vessel),6,10)
vessel_subject_id<-ifelse(substring(vessel_subject_id,5,5)==".",substring(vessel_subject_id,1,4),substring(vessel_subject_id,1,5))

vessel_together<-data.frame(cbind(vessel_subject_id,1))
colnames(vessel_together)<-c("id","haha")
index_data<-merge(x=vessel_together,y=pheno_together,by.x="id",all.x=T,sort=F)
dim(index_data)
```

Run the following code to selectgenes
```{r,message=FALSE,eval=FALSE}
#screening for smoking group
data_vessel_smoker<-data_vessel[,index_data$smoke==1]
index_data_smoker<-index_data[index_data$smoke==1,]

screening_index_smoking<-spearman_screen(index_data=index_data_smoker,data_vessel=data_vessel_smoker,size=40)
screening_index_smoking

#screening for non-smoking group
data_vessel_nsmoker<-data_vessel[,index_data$smoke==0]
index_data_nsmoker<-index_data[index_data$smoke==0,]

screening_index_nsmoking<-spearman_screen(index_data=index_data_nsmoker,data_vessel=data_vessel_nsmoker,size=40)
screening_index_nsmoking


sig_all_index<-test_screen(data_vessel,group=index_data$hyper,covariate=index_data$smoke)

index_utilized<-unique(c(screening_index_smoking,screening_index_nsmoking,sig_all_index))

saveRDS(index_utilized, file = "index_utilized.rds")

length(sig_all_index)

```

```{r,message=FALSE}
index_utilized<-readRDS(file = "index_utilized.rds")
```

We select 85 genes. Finally, let's construct the data for the following network learning
```{r,message=FALSE}
index<-as.numeric(as.matrix(index_data$rate,ncol=1))
data_vessel_order<-data_vessel[,order(index)] #order the gene matrix based on the order of the agent
x_original<-t(data_vessel_order)
data_observe<-x_original[,index_utilized] #focus on the selected 85 genes
agent<-index[order(index)] #order the agent
smk_cov<-index_data$smoke[order(index)]
```

# Train the varying coefficient model
Before constructing the network, each selected gene should be fitted by varying coefficient model. 
```{r,message=FALSE}
#further refine the data by deleting the data with agent too close to each other, this step is highly recommended in practice. Here we set the threshod=0.0001
data_observe<-data_observe[c(1,diff(agent,lag=1))>=0.0001,]
x_cov<-smk_cov[c(1,diff(agent,lag=1))>=0.0001]
agent<-agent[c(1,diff(agent,lag=1))>=0.0001]
```

Run the function “??” 

```{r,message=FALSE,eval=FALSE}
vc.fit<-function(agent=agent,data_observe=data_observe,x_cov=x_cov)
{
  t<-agent
#get fitted values of observations using varying coefficient kernel regression
data_fitted<-rep()
data_fitted_cov<-rep()
#bws<-0.01
for(i in 1:ncol(data_observe))
{
    bw <- npscoefbw(formula=data_observe[,i]~x_cov|t,betas=T,regtype="ll")
    bw <- npscoef(bw,betas=T,exdat=data.frame(x_cov=x_cov),ezdat=data.frame(t=t),regtype="ll")
    fitted<-coef(bw)[,1]
    data_fitted<-cbind(data_fitted,fitted)
    fitted_cov<-coef(bw)[,2]
    data_fitted_cov<-cbind(data_fitted_cov,fitted_cov)
}
return(list(data_fitted=data_fitted,data_fitted_cov=data_fitted_cov))
}

set.seed(4351)
data.vcfit<-vc.fit(agent=agent,data_observe=data_observe,x_cov=x_cov)
saveRDS(data.vcfit, file = "data.vcfit.rds")
```


```{r,message=FALSE}
data.vcfit<-readRDS(file = "data.vcfit.rds")
```

Visualize the fitting

```{r,message=FALSE}
id_sub<-8
{par(mfrow = c(1, 2))
#for smoking group
plot(x=agent[x_cov==1],y=data_observe[x_cov==1,id_sub],col="red",frame = FALSE,ylab="Expression",xlab="Agent")
lines(y=data.vcfit$data_fitted[,id_sub]+data.vcfit$data_fitted_cov[,id_sub],x=agent,col="red")

#for non-smoking group
plot(x=agent[x_cov==0],y=data_observe[x_cov==0,id_sub],col="blue",ylab="Expression",xlab="Agent")
lines(y=data.vcfit$data_fitted[,id_sub],x=agent,col="blue")}
```


# Generate base matrix for the network model

```{r,message=FALSE,eval=FALSE}
data_fitted<-data.vcfit$data_fitted
data_fitted_cov<-data.vcfit$data_fitted_cov

base.output<-base.construct(data_observe=data_observe,
    data_fitted=data_fitted, degree=3,len.knots=3,
data_fitted_cov=data_fitted_cov,
agent=agent,
x_cov=x_cov)

saveRDS(base.output, file = "base.output.rds")
  
```

```{r,message=FALSE}
base.output<-readRDS(file = "base.output.rds")
```

# Train the network model

```{r,message=FALSE,eval=FALSE}
X_big<-base.output$X_big
X_big_cov<-base.output$X_big_cov
X_big_int<-base.output$X_big_int
X_big_int_cov<-base.output$X_big_int_cov

network.output<-network.learn(data_observe=data_observe,
         x_cov=x_cov,
         X_big_int=X_big_int,
         X_big_int_cov=X_big_int_cov,
         agent=agent,
         degree=3,
         len.knots=3,
         cv=TRUE,
         nfolds=20,
         alpha=1)

summary(network.output$gene_whole_cov[10,])
data.list.t3<-list(data_observe=data_observe,
                   x_cov=x_cov,
                   agent=agent)
saveRDS(network.output, file = "network.output.rds")
saveRDS(data.list.t3, file = "data.list.t3.rds")
```

```{r,message=FALSE,eval=FALSE}
network.output<-readRDS(file = "network.output.rds")
```

```{r,message=FALSE}
library(igraph)
##need data
network.output<-readRDS(file = "network.output.rds")
data.list.t3<-readRDS(file = "data.list.t3.rds")
gene_exact_names<-read.csv(file="85_gene_name_information.csv")

data_observe<-data.list.t3$data_observe
x_cov<-data.list.t3$x_cov
agent<-data.list.t3$agent

self_size<-network.output$self_size
self_size_cov<-network.output$self_size_cov
self_size_all<-network.output$self_size_all
gene_whole_cov<-network.output$gene_whole_cov
gene_whole_all<-network.output$gene_whole_all
gene_whole<-network.output$gene_whole
```

```{r,message=FALSE}
##network for x=0 at i=250 after some tuning
par(mfrow = c(1, 1))
par(cex=1)
par(mar = c(0, 0, 0, 0), oma = c(6, 6, 3, 3))
tunned_value<-0.03
for (i in c(250))
{
    tunning_index<-apply(gene_whole_all,1,function (x) ifelse((abs(x[i+2]))>=tunned_value,1,0))
    gene_whole_all_tuned<-gene_whole_all[which(tunning_index==1),]
    tunning_index<-apply(gene_whole,1,function (x) ifelse((abs(x[i+2]))>=tunned_value,1,0))
    gene_whole_tuned<-gene_whole[which(tunning_index==1),]
    dim(gene_whole_tuned)
    unique(gene_whole_tuned[,2])
    table(gene_whole_tuned[,2])
    #geneate link data with 25th tissue
    links_gene<-gene_whole_tuned[,c(1,2,i+2)]
    colnames(links_gene)<-c("from","to","tissue_name")
    links_gene<-as.data.frame(links_gene)
    links_gene$edge_type<-ifelse(gene_whole_tuned[,i+2]>0,1,2)
    rownames(links_gene)<-rep(1:nrow(links_gene))
    
    self_index_all<-unique(c(gene_whole_all_tuned[,1],gene_whole_all_tuned[,2],gene_whole_tuned[,1],gene_whole_tuned[,2]))
    self_index<-unique(c(gene_whole_tuned[,1],gene_whole_tuned[,2]))
    #nodes_gene<-data.frame(cbind(self_index,self_size[self_index,(i+1)]))
    nodes_gene<-data.frame(self_size[self_size[,1] %in% self_index,c(1,(i+1))])
    colnames(nodes_gene)<-c("id","self_weight")
    #nodes_gene<-data.frame(self_size[,c((i+1))])
    #colnames(nodes_gene)<-c("self_weight")
    nodes_gene$node_type<-ifelse(nodes_gene$id %in% c(39,40,48,64,76),1, ifelse(nodes_gene$id %in% (self_index),2,3))
    
    #nodes_gene$node_type<-ifelse(nodes_gene$id %in% self_index,3,4)
    net <- graph.data.frame(links_gene, nodes_gene, directed=T) 
    class(net)
    net 
    
    # It's easy to access nodes, edges, and their attributes:
    E(net)
    V(net)$self_weight
    E(net)$tissue_name
    V(net) #total 47 clusters included in the network
    
    # Now we should be able to do this:
    #plot(net, edge.arrow.size=.2,vertex.label.family="Arial Black" )
    
    #adjust size of nodes and width of edges
    colrs_nodes <- c("red","orange","gray")
    V(net)$color <- colrs_nodes[V(net)$node_type]
    colrs_edge<-c("tomato","gray40")
    E(net)$color <- colrs_edge[E(net)$edge_type]
    V(net)$size <- (V(net)$self_weight)^0.7*4
    E(net)$width <- (abs(E(net)$tissue_name))^0.4*7
    
    #change coordinate to separate the nodes
    
    #set.seed(41414343)
    net.bg<-net
    set.seed(4151320)
    l<-layout_nicely(net.bg)
    l <- layout.norm(l, ymin=-1, ymax=1, xmin=-1, xmax=1)
    l[20,]<-c(-0.9,0.1)
    l[16,]<-c(0.7,-0.5)
    l[46,]<-c(-0.7,-0.7)
    l[8,]<-c(0.7,-0.7)
    #l<-layout_in_circle(net.bg)
    #l<-layout_as_star(net.bg,center=V(net)[1:3])
    #l<-layout_as_bipartite(net.bg)
    #l<-layout_on_grid(net.bg)
    #l<-layout_on_sphere(net.bg)
    #l<-layout_with_sugiyama(net.bg)
    #l<-layout_with_dh(net.bg)
    # Normalize them so that they are in the -1, 1 interval:
    #l <- layout.norm(l, ymin=-1, ymax=1, xmin=-1, xmax=1)
    gene_names_used<-as.character(gene_exact_names$gene.name[nodes_gene$id])
    #gene_names_used[which(gene_names_used=="CTSD\xca")]<-"CTSD"
    plot.igraph(net.bg, rescale=F, layout=l,edge.arrow.size=.1, edge.curved=.2,vertex.label=gene_names_used,
                vertex.label.color="blue",vertex.label.cex=0.7,mark.border=NA)
}
```

```{r,message=FALSE}
{par(mfrow = c(1, 3))
tunned_value<-0.01
for (i in c(10,100,250))
{
    tunning_index<-apply(gene_whole_cov,1,function (x) ifelse((abs(x[i+2]))>=tunned_value,1,0))
    gene_whole_cov_tuned<-gene_whole_cov[which(tunning_index==1),]
    dim(gene_whole_cov_tuned)
    unique(gene_whole_cov_tuned[,2])
    table(gene_whole_cov_tuned[,2])
    #geneate link data with 25th tissue
    links_gene<-gene_whole_cov_tuned[,c(1,2,i+2)]
    colnames(links_gene)<-c("from","to","tissue_name")
    links_gene<-as.data.frame(links_gene)
    links_gene$edge_type<-ifelse(gene_whole_cov_tuned[,i+2]>0,1,2)
    rownames(links_gene)<-rep(1:nrow(links_gene))
    
    #self_index_all<-unique(c(gene_whole_tuned[,1],gene_whole_tuned[,2],gene_whole_cov_tuned[,1],gene_whole_cov_tuned[,2]))
    self_index_cov<-unique(c(gene_whole_cov_tuned[,1],gene_whole_cov_tuned[,2]))
    self_index_all<-self_index_cov
    #nodes_gene<-data.frame(cbind(self_index,self_size[self_index,(i+1)]))
    nodes_gene<-abs(data.frame(self_size_cov[self_size_cov[,1] %in% self_index_all,c(1,(i+1))]))
    colnames(nodes_gene)<-c("id","self_weight")
    #nodes_gene<-data.frame(self_size[,c((i+1))])
    #colnames(nodes_gene)<-c("self_weight")
    nodes_gene$node_type<-ifelse(nodes_gene$id %in% (self_index_cov),3,4)
    net <- graph.data.frame(links_gene, nodes_gene, directed=T) 
    class(net)
    net 
    
    # It's easy to access nodes, edges, and their attributes:
    E(net)
    V(net)$self_weight
    E(net)$tissue_name
    V(net) #total 34 clusters included in the network
    
    # Now we should be able to do this:
    #plot(net, edge.arrow.size=.2,vertex.label.family="Arial Black" )
    
    #adjust size of nodes and width of edges
    colrs_nodes <- c("red", "green","orange","gray")
    V(net)$color <- colrs_nodes[V(net)$node_type]
    colrs_edge<-c("tomato","gray40")
    E(net)$color <- colrs_edge[E(net)$edge_type]
    V(net)$size <- (V(net)$self_weight)^0.7*4
    E(net)$width <- (abs(E(net)$tissue_name))^0.4*7
    net.bg<-net
    set.seed(4151320)
    l<-layout_nicely(net.bg)
    l <- layout.norm(l, ymin=-1, ymax=1, xmin=-1, xmax=1)
    #l[1,2]<-0.3
    #l[3,2]<-0.7
    #l[25,2]<--0.5
    #l[23,2]<-0
    #l[14,2]<--0.6
    #l[8,]<-c(-0.9,-0.35)
    #l[16,1]<-0.4
    #l[15,1]<--0.2
    #l<-layout_in_circle(net.bg)
    #l<-layout_as_star(net.bg,center=V(net)[1:3])
    #l<-layout_as_bipartite(net.bg)
    #l<-layout_on_grid(net.bg)
    #l<-layout_on_sphere(net.bg)
    #l<-layout_with_sugiyama(net.bg)
    #l<-layout_with_dh(net.bg)
    # Normalize them so that they are in the -1, 1 interval:
    l <- layout.norm(l, ymin=-1, ymax=1, xmin=-1, xmax=1)
    gene_names_used<-as.character(gene_exact_names$gene.name[nodes_gene$id])
    gene_names_used[which(gene_names_used=="CTSD\xca")]<-"CTSD"
    plot.igraph(net.bg, rescale=F, layout=l*0.9,edge.arrow.size=.12, edge.curved=.2,vertex.label=gene_names_used,
                vertex.label.color="blue",vertex.label.cex=0.7,mark.border=NA)
}
}

```




