---
title: "Tutorial 3: Visualize the results  "
author: "Chixiang Chen"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{DRDNetPro}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '/Users/chixiang.chen/OneDrive - University of Maryland School of Medicine/phd_in_psu/research/Prof._Wu/Network/tissue/blood_vessel/')
```

**DRDNetPro** is a statistical learning algorithm for recovering disease risk-associated pseudo-dynamic networks (DRDNet) from steady-state data. It incorporates risk prediction model, a varying coefficient model, multiple ordinary differential equations, and group lasso estimation to learn a series of networks.This tutorial will provide detailed information for the third step of implementing the [DRDNet](https://academic.oup.com/bioinformatics/article-abstract/38/9/2481/6537533). 

If the user hasn't gone through the Tutorial 1, please visit [Tutorial 2](https://chencxxy28.github.io/DRDNetPro/articles/web/Tutorial2.html) for the second step of training the network model. 

To begin with, load the following packages

```{r}
library(igraph)

```

The data we need to use in this tutorial is [network.output], [data.list.t3], and [85_gene_name_information]
```{r,message=FALSE}
##need data
network.output<-readRDS(file = "network.output.rds")
data.list.t3<-readRDS(file = "data.list.t3.rds")
gene_exact_names<-read.csv(file="85_gene_name_information.csv")
```

Extract detailed data about gene expression data, covariate vector, imputed agent, self-node size for baseline, self-node size for covariate effect, overall self-node size, gene interaction effects from the baseline, gene interaction effects from the covariate, and overall gene interaction effects.
```{r,message=FALSE}
data_observe<-data.list.t3$data_observe
x_cov<-data.list.t3$x_cov
agent<-data.list.t3$agent

self_size<-network.output$self_size
self_size_cov<-network.output$self_size_cov
self_size_all<-network.output$self_size_all

gene_whole_cov<-network.output$gene_whole_cov
gene_whole_all<-network.output$gene_whole_all
gene_whole<-network.output$gene_whole
```

To visualize the pseudo dynamic network given an agent value, a useful tool is `igraph` package. It allows the self-designed network structure. More detailed tutorial about igraph can be found in [igraph](https://kateto.net/networks-r-igraph). The following is a demo code for visualizing the recovered baseline network, where the disease risk is changing from 0 (low risk) to 0.86 (high risk). 
```{r,message=FALSE}
##network for x=0 at i=250 after some tuning
par(mfrow = c(1, 1))
par(cex=1)
par(mar = c(0, 0, 0, 0), oma = c(6, 6, 3, 3))
tunned_value<-0.03
for (i in c(250))
{
    tunning_index<-apply(gene_whole_all,1,function (x) ifelse((abs(x[i+2]))>=tunned_value,1,0))
    gene_whole_all_tuned<-gene_whole_all[which(tunning_index==1),]
    tunning_index<-apply(gene_whole,1,function (x) ifelse((abs(x[i+2]))>=tunned_value,1,0))
    gene_whole_tuned<-gene_whole[which(tunning_index==1),]
    dim(gene_whole_tuned)
    unique(gene_whole_tuned[,2])
    table(gene_whole_tuned[,2])
    #geneate link data with 25th tissue
    links_gene<-gene_whole_tuned[,c(1,2,i+2)]
    colnames(links_gene)<-c("from","to","tissue_name")
    links_gene<-as.data.frame(links_gene)
    links_gene$edge_type<-ifelse(gene_whole_tuned[,i+2]>0,1,2)
    rownames(links_gene)<-rep(1:nrow(links_gene))
    
    self_index_all<-unique(c(gene_whole_all_tuned[,1],gene_whole_all_tuned[,2],gene_whole_tuned[,1],gene_whole_tuned[,2]))
    self_index<-unique(c(gene_whole_tuned[,1],gene_whole_tuned[,2]))
    #nodes_gene<-data.frame(cbind(self_index,self_size[self_index,(i+1)]))
    nodes_gene<-data.frame(self_size[self_size[,1] %in% self_index,c(1,(i+1))])
    colnames(nodes_gene)<-c("id","self_weight")
    #nodes_gene<-data.frame(self_size[,c((i+1))])
    #colnames(nodes_gene)<-c("self_weight")
    nodes_gene$node_type<-ifelse(nodes_gene$id %in% c(39,40,48,64,76),1, ifelse(nodes_gene$id %in% (self_index),2,3))
    
    #nodes_gene$node_type<-ifelse(nodes_gene$id %in% self_index,3,4)
    net <- graph.data.frame(links_gene, nodes_gene, directed=T) 
    class(net)
    net 
    
    # It's easy to access nodes, edges, and their attributes:
    E(net)
    V(net)$self_weight
    E(net)$tissue_name
    V(net) #total 47 clusters included in the network
    
    # Now we should be able to do this:
    #plot(net, edge.arrow.size=.2,vertex.label.family="Arial Black" )
    
    #adjust size of nodes and width of edges
    colrs_nodes <- c("red","orange","gray")
    V(net)$color <- colrs_nodes[V(net)$node_type]
    colrs_edge<-c("tomato","gray40")
    E(net)$color <- colrs_edge[E(net)$edge_type]
    V(net)$size <- (V(net)$self_weight)^0.7*4
    E(net)$width <- (abs(E(net)$tissue_name))^0.4*7
    
    #change coordinate to separate the nodes
    
    #set.seed(41414343)
    net.bg<-net
    set.seed(4151320)
    l<-layout_nicely(net.bg)
    l <- layout.norm(l, ymin=-1, ymax=1, xmin=-1, xmax=1)
    l[20,]<-c(-0.9,0.1)
    l[16,]<-c(0.7,-0.5)
    l[46,]<-c(-0.7,-0.7)
    l[8,]<-c(0.7,-0.7)
    #l<-layout_in_circle(net.bg)
    #l<-layout_as_star(net.bg,center=V(net)[1:3])
    #l<-layout_as_bipartite(net.bg)
    #l<-layout_on_grid(net.bg)
    #l<-layout_on_sphere(net.bg)
    #l<-layout_with_sugiyama(net.bg)
    #l<-layout_with_dh(net.bg)
    # Normalize them so that they are in the -1, 1 interval:
    #l <- layout.norm(l, ymin=-1, ymax=1, xmin=-1, xmax=1)
    gene_names_used<-as.character(gene_exact_names$gene.name[nodes_gene$id])
    #gene_names_used[which(gene_names_used=="CTSD\xca")]<-"CTSD"
    plot.igraph(net.bg, rescale=F, layout=l,edge.arrow.size=.1, edge.curved=.2,vertex.label=gene_names_used,
                vertex.label.color="blue",vertex.label.cex=0.7,mark.border=NA)
}
```


The following is a demo code for visualizing three recovered covariate-effect networks, where the disease risk is changing from 0 (no risk) to 0.10 (low risk), from 0 (no risk) to 0.42 (moderate risk), and from 0 (no risk) to 0.86 (high risk). 

```{r,message=FALSE}
{par(mfrow = c(1, 3))
tunned_value<-0.01
for (i in c(10,100,250))
{
    tunning_index<-apply(gene_whole_cov,1,function (x) ifelse((abs(x[i+2]))>=tunned_value,1,0))
    gene_whole_cov_tuned<-gene_whole_cov[which(tunning_index==1),]
    dim(gene_whole_cov_tuned)
    unique(gene_whole_cov_tuned[,2])
    table(gene_whole_cov_tuned[,2])
    #geneate link data with 25th tissue
    links_gene<-gene_whole_cov_tuned[,c(1,2,i+2)]
    colnames(links_gene)<-c("from","to","tissue_name")
    links_gene<-as.data.frame(links_gene)
    links_gene$edge_type<-ifelse(gene_whole_cov_tuned[,i+2]>0,1,2)
    rownames(links_gene)<-rep(1:nrow(links_gene))
    
    #self_index_all<-unique(c(gene_whole_tuned[,1],gene_whole_tuned[,2],gene_whole_cov_tuned[,1],gene_whole_cov_tuned[,2]))
    self_index_cov<-unique(c(gene_whole_cov_tuned[,1],gene_whole_cov_tuned[,2]))
    self_index_all<-self_index_cov
    #nodes_gene<-data.frame(cbind(self_index,self_size[self_index,(i+1)]))
    nodes_gene<-abs(data.frame(self_size_cov[self_size_cov[,1] %in% self_index_all,c(1,(i+1))]))
    colnames(nodes_gene)<-c("id","self_weight")
    #nodes_gene<-data.frame(self_size[,c((i+1))])
    #colnames(nodes_gene)<-c("self_weight")
    nodes_gene$node_type<-ifelse(nodes_gene$id %in% (self_index_cov),3,4)
    net <- graph.data.frame(links_gene, nodes_gene, directed=T) 
    class(net)
    net 
    
    # It's easy to access nodes, edges, and their attributes:
    E(net)
    V(net)$self_weight
    E(net)$tissue_name
    V(net) #total 34 clusters included in the network
    
    # Now we should be able to do this:
    #plot(net, edge.arrow.size=.2,vertex.label.family="Arial Black" )
    
    #adjust size of nodes and width of edges
    colrs_nodes <- c("red", "green","orange","gray")
    V(net)$color <- colrs_nodes[V(net)$node_type]
    colrs_edge<-c("tomato","gray40")
    E(net)$color <- colrs_edge[E(net)$edge_type]
    V(net)$size <- (V(net)$self_weight)^0.7*4
    E(net)$width <- (abs(E(net)$tissue_name))^0.4*7
    net.bg<-net
    set.seed(4151320)
    l<-layout_nicely(net.bg)
    l <- layout.norm(l, ymin=-1, ymax=1, xmin=-1, xmax=1)
    #l[1,2]<-0.3
    #l[3,2]<-0.7
    #l[25,2]<--0.5
    #l[23,2]<-0
    #l[14,2]<--0.6
    #l[8,]<-c(-0.9,-0.35)
    #l[16,1]<-0.4
    #l[15,1]<--0.2
    #l<-layout_in_circle(net.bg)
    #l<-layout_as_star(net.bg,center=V(net)[1:3])
    #l<-layout_as_bipartite(net.bg)
    #l<-layout_on_grid(net.bg)
    #l<-layout_on_sphere(net.bg)
    #l<-layout_with_sugiyama(net.bg)
    #l<-layout_with_dh(net.bg)
    # Normalize them so that they are in the -1, 1 interval:
    l <- layout.norm(l, ymin=-1, ymax=1, xmin=-1, xmax=1)
    gene_names_used<-as.character(gene_exact_names$gene.name[nodes_gene$id])
    gene_names_used[which(gene_names_used=="CTSD\xca")]<-"CTSD"
    plot.igraph(net.bg, rescale=F, layout=l*0.9,edge.arrow.size=.12, edge.curved=.2,vertex.label=gene_names_used,
                vertex.label.color="blue",vertex.label.cex=0.7,mark.border=NA)
}
}

```




